MetaMask/metamask-extension
# Remove global transaction state from send flow

## Explanation

As part of exploring possible causes of #11677 and came to the conclusion that, even though I could not reproduce the bug, that it was possible for the send slice to get polluted with bad data. There were two main problems with the structure of the send slice:

1. Upon every update to the transaction state we generated the txParams and kept them in the slice as well. txParams are a derivative of the other state tree values but under certain scenarios it was plausible that the txParams in state would become stale and not be an accurate representation of the rest of the state tree. To resolve this issue I removed the derivative state from the slice in #14701 
2. The transaction state in the send slice is global to the slice, meaning that the slice is "initiated" when navigating to the send page and the send page would clear the send slice state after the send flow ended...but it was plausible that a new send flow could be kicked off in some way while there was lingering state from the previous send in the tree. This would cause an issue where the transaction would get corrupted with the previous transaction's intent. 

This PR aims to alleviate the second point.

Send flow before this change:
1. User clicks send in the extension main screen.
2. If the user was on an asset page or clicked send from a token  asset, we'd dispatch `updateSendAsset` with the token details before routing the user to the send page.
3. The send page had a `useEffect` that would dispatch `initializeSendState` this would do things like fetch gas price estimations, set the validity of the state, etc. 
4. The transaction values such as recipient, amount, from and to were stored in a global state tree.
5. Upon leaving the send flow the send page would dispatch the `resetSendState` action that would set the tree back to default.

The same basic flow would happen upon editing a transaction, except that in addition to setting some global values the editTransaction method would be dispatched that also set the details of the edited transaction into the global state.

Send flow after this change:
1. user clicks send in the extension main screen
2. we dispatch `startNewTransaction` and include asset details in that call.
3. The send slice clears all previous draft transactions from state, which are now stored in an object in state that is no longer global. 
4. The send slice generates a new UUID and sets as the `currentTransactionUUID` in state, then creates a blank transaction in the `draftTransactions` object that is keyed by UUID.
5. The `initializeSendState` thunk is dispatched.
6. Once complete the user is routed to the send page. 
7. Upon leaving the send flow the state is cleared.

if editing a transaction a new 'editExistingTransaction' method is dispatched with the asset type and id of the transaction, but the rest of the flow is the same. 

### How does this help? 
1. All transaction values are kept inside an object that is keyed by a non deterministic UUID generated at the time the draft is created.  
2. When a new draft is created, all previous drafts are discarded as an additional fail safe to make sure that the new transaction has no chance to be confused with an older, stale one. 
3. We no longer manipulate the send state ahead of kicking off the send flow which makes the send flow less reliant upon side effects and makes it idempotent, in that it is less likely to have a different outcome based on the order of operations.

### Related Changes
1. A new 'getCurrentDraftTransaction' selector was added as a tool for all other selectors. Most selectors now call this selector to get either an empty object or the current draftTransaciton.
2. A lot of selectors have been updated to reflect the possibility that a draftTransaction does not exist. 
3. For some reason in some of our confirmation components we use the send slice selectors, I did not change this but we can be reasonably assured now that they aren't grabbing unrelated send state. 
4. The send slice has been reorganized a great deal to try and keep things alphabetical where possible. It is not always possible, especially in the selectors area, due to things being used before they are declared. 
5. More types and safer types have been added.
6. A number of helper methods were extracted from the send slice file to a helpers.js file in the send slice folder. This is to make the send slice less difficult to read.

## More Information
Fixes #11677 

## Screenshots/Screencaps



### Before & After
The functionality and flow of the send page is unchanged, no E2E tests needed to be modified to make this work. 

## Manual Testing Steps
1. Test sending ETH, Tokens, Collectibles.
2. Test changing accounts, networks, and assets on the send flow.
3. Test changing the send amount in the send flow.
4. Test editing transactions.
5. Ensure that every transaction you create through the send flow results in a confirmation with matching details.
7. Check ENS, copy and paste, and manual address entry in the recipient fielld.
8. Test adding nicknames to addresses and see that the nickname is reflected in the send flow. 
9. Any other details related to the send flow.

## Pre-Merge Checklist
